// cftools-sdk/lib/internal/got/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toHitZones = void 0;
function toHitZones(zones) {
    return {
        torso: (zones === null || zones === void 0 ? void 0 : zones.torso) || 0,
        rightLeg: (zones === null || zones === void 0 ? void 0 : zones.rightleg) || 0,
        rightHand: (zones === null || zones === void 0 ? void 0 : zones.righthand) || 0,
        rightArm: (zones === null || zones === void 0 ? void 0 : zones.rightarm) || 0,
        rightFoot: (zones === null || zones === void 0 ? void 0 : zones.rightfoot) || 0,
        leftLeg: (zones === null || zones === void 0 ? void 0 : zones.leftleg) || 0,
        head: (zones === null || zones === void 0 ? void 0 : zones.head) || 0,
        leftFoot: (zones === null || zones === void 0 ? void 0 : zones.leftfoot) || 0,
        leftArm: (zones === null || zones === void 0 ? void 0 : zones.leftarm) || 0,
        leftHand: (zones === null || zones === void 0 ? void 0 : zones.lefthand) || 0,
        brain: (zones === null || zones === void 0 ? void 0 : zones.brain) || 0,
    };
}
exports.toHitZones = toHitZones;
